test/units/TEST-07-PID1.protect-control-groups.sh

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: LGPL-2.1-or-later
+# shellcheck disable=SC2016
+set -eux
+set -o pipefail
+
+# shellcheck source=test/units/test-control.sh
+. "$(dirname "$0")"/test-control.sh
+# shellcheck source=test/units/util.sh
+. "$(dirname "$0")"/util.sh
+
+SLICE="system.slice"
+UNIT="test-07-protect-control-groups"
+
+READ_ONLY_MOUNT_FLAG="ro"
+READ_WRITE_MOUNT_FLAG="rw"
+
+at_exit() {
+    set +e
+
+    systemctl stop "$UNIT"
+    systemctl reset-failed
+}
+
+trap at_exit EXIT
+
+ROOT_CGROUP_NS=$(readlink /proc/self/ns/cgroup)
+
+test_basic() {
+    local protect_control_groups_ex="$1"
+    local protect_control_groups="$2"
+    local in_cgroup_ns="$3"
+    local mount_flag="$4"
+
+    if [[ $in_cgroup_ns == true ]]; then
+      local ns_cmp_op="!="
+      local unit_cgroup="0::/"
+      local memory_pressure_watch="/sys/fs/cgroup/memory.pressure"
+    else
+      local ns_cmp_op="=="
+      local unit_cgroup="0::/$SLICE/$UNIT.service"
+      local memory_pressure_watch="/sys/fs/cgroup/$SLICE/$UNIT.service/memory.pressure"
+    fi
+
+    # Compare cgroup namespace to root namespace
+    systemd-run -p "ProtectControlGroupsEx=$protect_control_groups_ex" --slice "$SLICE" --unit "$UNIT" --wait \
+            bash -xec "test \"\$(readlink /proc/self/ns/cgroup)\" $ns_cmp_op \"$ROOT_CGROUP_NS\""
+    # Verify unit cgroup
+    systemd-run -p "ProtectControlGroupsEx=$protect_control_groups_ex" --slice "$SLICE" --unit "$UNIT" --wait \
+            bash -xec "test \"\$(cat /proc/self/cgroup)\" == \"$unit_cgroup\""
+    # Verify memory pressure watch points to correct file
+    systemd-run -p "ProtectControlGroupsEx=$protect_control_groups_ex" --slice "$SLICE" --unit "$UNIT" --wait \
+            bash -xec "test \"\$MEMORY_PRESSURE_WATCH\" == \"$memory_pressure_watch\""
```

Comment:
I am not sure why this fails, but `$MEMORY_PRESSURE_WATCH` is not set in some cases.
I recommend to set `-p Environment=SYSTEMD_LOG_LEVEL=debug`.

Reply:
I think the issue here hopefully is fairly simple...I wasn't setting MemoryPressureWatch=on and instead relying on a distro's DefaultMemoryPressureWatch. Now I set it, it should be okay.

It could be theoretically be possible a kernel doesn't support memory pressure or /sys/fs/cgroup isn't mounted with nsdelegate but all CI should run on kernels >= 5.2 so this should be fine, I think? I guess we'll found it when CI runs

Reply:
Hmmm looks like C9 upstream was still failing when setting MemoryPressureWatch=on - now the whole unit was failing. My guess is C9 upstream build environment is customized s.t. they don't mount /sys/fs/cgroup with nsdelegate. So adding checks to verify kernel support and memory pressure delegation exist